### DISCUSSION TITLE
Different behaviour on custom type with String() method

### DISCUSSION THREAD

**Discussion by pgundlach** [dop251/goja#619] [Q&A] [answered]
When I comment the `func (mt mytype) String() string` method below, I get a different assignment in the map. See the result in the `Check()` method.

Is there a way around this without redefining everything from mytype? I cannot change the underlying code (which has a `String() string` method) and mytype has lots of values attached to it.

```
package main

import (
	"fmt"
	"os"

	"github.com/dop251/goja"
)

type keytype int

const key keytype = 123

func (mt keytype) String() string {
	return "abc"
}

type document struct {
	Items map[keytype]any
}

func (d *document) Check() {
	// prints 123 if String method is commented
	// prints 0 if String method is defined
	for k := range d.Items {
		fmt.Printf("%d\n", k)
	}
}

func main() {
	runtime := goja.New()
	runtime.Set("newdoc", func() *document { return &document{Items: make(map[keytype]any)} })
	runtime.Set("key", key)
	_, err := runtime.RunString(`
d = newdoc();
d.Items[key] = "bar";
d.Check();
`)
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}
}
```

**Comment by dop251** [comment, marked as answer]
This one is a bit tricky... Properties in ECMAScript can only be strings or Symbols, so the key must be converted into string before used as a property name, and if a value is a `fmt.Stringer` then the `String()` method is used for that.

If you can change the javascript code, you can work around that by doing `d.Items[Number(key)] = "bar"`.

**Reply by pgundlach** [reply]
The `Number()` works fine. This leads me to this trick, which work as well:

```
runtime.Set("key", runtime.ToValue(key).ToNumber())
```

in the code above. I hope there is no drawback (performance issues are not that important)

Thank you very much again!
